### ISSUE TITLE
Support map attribute values

### ISSUE THREAD

**Issue by jack-berg** [open-telemetry/opentelemetry-java#3542]
**Is your feature request related to a problem? Please describe.**
The protos support map attribute values via the [KeyValueList](https://github.com/open-telemetry/opentelemetry-proto/blame/main/opentelemetry/proto/common/v1/common.proto#L36) type. The otel java API / SDK should support map attribute values. 

**Describe the solution you'd like**
Add map APIs / implementations to the attribute package in `io.opentelemetry.api.common`. 

**Describe alternatives you've considered**
N/A.

**Additional context**
N/A.

**Comment by Oberon00** [comment]
That protocol feature is only meant for use with logs (which are not supported at all currently). The "normal" attributes that the tracing API should support are described in the specification here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attributes

You cannot just start allowing more attribute types in an 1.x version, as that would potentially break third party exporters and span processors.

**Comment by jack-berg** [comment]
> That protocol feature is only meant for use with logs

Interesting. Looking for evidence of this in the protos, I see a comment on [trace.proto](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L161-L169) which references different attribute types, which excludes maps, but also excludes arrays which are definitely supported. No mention of unsupported types in the [metric.proto](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L323), or [resource.proto](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/resource/v1/resource.proto#L29). 

So is the idea that the next major version of the spec may expand map attribute types to other signals (metrics / logs), but for now apis / sdks are not expected to support those types despite availability in the protos?

**Comment by jkwatson** [comment]
I would prefer not to complicate our APIs with support for this. @jack-berg do you have a real-life use-case for this at this point?

**Comment by Oberon00** [comment]
@jack-berg You need to do some archeology to find that 😃 See PR description here: https://github.com/open-telemetry/opentelemetry-proto/pull/157

Anyway, the reference for what the API should support is the specification (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md?rgh-link-date=2021-08-24T16%3A51%3A41Z#attributes), not what some protocol that we support supports (even though I concede that OTLP is our "primary" protocol).

I know of at least one time this feature was discussed in the spec here: https://github.com/open-telemetry/opentelemetry-specification/pull/596

**Comment by jack-berg** [comment]
@jkwatson I opened this issue mostly to prompt a discussion for I what I interpreted was a discrepancy between the spec and the java implementation. 

Upon further clarification, especially from this [comment](https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-674243387), it appears that there is an acknowledged difference between the attributes in the OTLP proto and in the API. (Thanks for helping clarify that @Oberon00!)

I think we should close this issue until the [spec issue](https://github.com/open-telemetry/opentelemetry-specification/issues/376) is resolved.
